File: QuickStarts/VB/Stoplight/Tests/StopLightTests/StoplightFixture.vb
'===============================================================================
' Microsoft patterns & practices
' Unity Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports StopLight.Logic
Imports StopLightTests.TestDoubles

Namespace StopLightTests
	''' <summary>
	''' Summary description for StopLightFixture
	''' </summary>
  <TestClass()> _
  Public Class StoplightFixture

    Dim newColor As StoplightColors
    Dim eventFired As Boolean

    Private Sub SetColorAndEventFired(ByVal sender As Object, ByVal e As LightChangedEventArgs)
      eventFired = True
      newColor = e.CurrentColor
    End Sub

    <TestMethod()> _
    Public Sub ShouldInitializeToGreen()
      Dim light As New StopLight.Logic.Stoplight()
      Assert.AreEqual(StopLight.Logic.Stoplight.Green, light.CurrentColor)
    End Sub

    <TestMethod()> _
    Public Sub ShouldTransitionGreenToYellow()
      Dim light As New StopLight.Logic.Stoplight()
      light.[Next]()

      Assert.AreEqual(StopLight.Logic.Stoplight.Yellow, light.CurrentColor)
    End Sub

    <TestMethod()> _
    Public Sub ShouldTransitionYellowToRed()
      Dim light As New StopLight.Logic.Stoplight()
      light.[Next]()
      light.[Next]()

      Assert.AreEqual(StopLight.Logic.Stoplight.Red, light.CurrentColor)
    End Sub

    <TestMethod()> _
    Public Sub ShouldTransitionRedToGreen()
      Dim light As New StopLight.Logic.Stoplight()
      light.[Next]()
      light.[Next]()
      light.[Next]()

      Assert.AreEqual(StopLight.Logic.Stoplight.Green, light.CurrentColor)
    End Sub

    <TestMethod()> _
    Public Sub ShouldRaiseChangedEventOnTransition()
      Dim light As New StopLight.Logic.Stoplight()
      eventFired = False
      newColor = StoplightColors.Green
      AddHandler light.Changed, AddressOf SetColorAndEventFired
      light.[Next]()
      Assert.IsTrue(eventFired)
      Assert.AreEqual(StopLight.Logic.Stoplight.Yellow, newColor)
    End Sub

    <TestMethod()> _
    Public Sub ShouldLogMessageOnChange()
      Dim light As New StopLight.Logic.Stoplight()
      Dim logger As New MockLogger()
      light.Logger = logger
      light.[Next]()
      Assert.IsNotNull(logger.LastMessage)
      Assert.IsTrue(logger.LastMessage.StartsWith("LIGHT CHANGED TO"))
    End Sub
  End Class
End Namespace

File: quickstarts/CS/Stoplight/Tests/StopLightTests/StoplightFixture.cs
// Copyright (c) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.

using Microsoft.VisualStudio.TestTools.UnitTesting;
using StopLight.Logic;
using StopLightTests.TestDoubles;

namespace StopLightTests
{
	/// <summary>
	/// Summary description for StopLightFixture
	/// </summary>
	[TestClass]
	public class StoplightFixture
	{
		[TestMethod]
		public void ShouldInitializeToGreen()
		{
			Stoplight light = new Stoplight();
			Assert.AreEqual(Stoplight.Green, light.CurrentColor);
		}

		[TestMethod]
		public void ShouldTransitionGreenToYellow()
		{
			Stoplight light = new Stoplight();
			light.Next();

			Assert.AreEqual(Stoplight.Yellow, light.CurrentColor);
		}

		[TestMethod]
		public void ShouldTransitionYellowToRed()
		{
			Stoplight light = new Stoplight();
			light.Next();
			light.Next();

			Assert.AreEqual(Stoplight.Red, light.CurrentColor);
		}

		[TestMethod]
		public void ShouldTransitionRedToGreen()
		{
			Stoplight light = new Stoplight();
			light.Next();
			light.Next();
			light.Next();

			Assert.AreEqual(Stoplight.Green, light.CurrentColor);
		}

		[TestMethod]
		public void ShouldRaiseChangedEventOnTransition()
		{
			bool eventFired = false;
			StoplightColors newColor = StoplightColors.Green;
			Stoplight light = new Stoplight();
			light.Changed += delegate(object sender, LightChangedEventArgs e)
			                 {
			                 	eventFired = true;
			                 	newColor = e.CurrentColor;
			                 };
			
			light.Next();
			Assert.IsTrue(eventFired);
			Assert.AreEqual(Stoplight.Yellow, newColor);
		}

		[TestMethod]
		public void ShouldLogMessageOnChange()
		{
			Stoplight light = new Stoplight();
			MockLogger logger = new MockLogger();
			light.Logger = logger;

			light.Next();

			Assert.IsNotNull(logger.LastMessage);
			Assert.IsTrue(logger.LastMessage.StartsWith("LIGHT CHANGED TO"));
		}
	}
}
